src/OrchardCore/OrchardCore.Mvc.Core/Utilities/ControllerExtensions.cs

```diff
@@ -8,11 +8,18 @@ public static class ControllerExtensions
         /// Returns the proper actionresult for unauthorized or unauthenticated users.
         /// Will return a forbid when the user is authenticated.
         /// Will return a challenge when the user is not authenticated.
+        /// If a authentication scheme is specified, will return a challenge to it.
         /// </summary>
         /// <param name="controller"></param>
+        /// <param name="authenticationScheme"></param>
         /// <returns>The proper actionresult based upon if the user is authenticated</returns>
-        public static ActionResult ChallengeOrForbid(this Controller controller)
+        public static ActionResult ChallengeOrForbid(this Controller controller, string authenticationScheme = null)
         {
+            if (!string.IsNullOrEmpty(authenticationScheme))
+            {
+                return controller.User.Identity.IsAuthenticated ? (ActionResult)controller.Forbid() : (ActionResult)controller.Challenge(new string[] { authenticationScheme });
```

Comment:
We should also call `Forbid(new string[] { authenticationScheme })` instead of just `Forbid()`.

Reply:
I think there's an overload that just takes a string value to avoid creating `new string[]` unless the api has changed.

Reply:
There is no method with only a string: https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.controllerbase.challenge?view=aspnetcore-3.1

Reply:
`Forbid(new string[] { authenticationScheme })` added.

Reply:
Just updated the extensions to match what is used in `ControllerBase`